execution-env.md

```diff
@@ -1,80 +1,119 @@
+1. Validated the pipeline configuration and all associated objects.
+1. Determined the set of data resources that should be accessible to the shaders, and in what access modes (e.g. read-only, read-write).
+1. Determined the computing resources to be used to execute the pipeline.
+1. Created a pipeline ExecutionContext consisting of the computing and data resources that are valid to be used by the pipeline, and in particular by the shaders in the pipeline. 
+   Each data resource is associated with valid access modes: read-only resources may only be read; only writable resources may be written.
+1. Executes the pipeline, subject to the security property that follows.
 
 A WebGPU implementation is _secure_ if, for any application _App_ running on the implementation, accesses explicitly granted to App are the only means by which:
 
-
-
 *   _App_ may sense state outside its ExecutionContext
 *   _App_ may cause mutation of state outside its ExecutionContext
 *   _App_ may cause or sense I/O outside its ExecutionContext
 
 An access is considered explicitly granted only if:
 
-
-
 *   It is a read access to a resource attached for reading to the ExecutionContext
 *   It is a write access to a resource attached for writing to the ExecutionContext
 
+Typically a WebGPU implementations can make sure that the _App_ can only access resources provided in the ExecutionContext by validating the code in the shader module.
+However this only guarantees that the correct resources are accessed, not that they are accessed in bounds.
+This is an issue because resource accesses are often lowered to pointer arithmetic so out of bounds accesses could result in accesses outside the ExecutionContext.
+For this reason out of bound accesses to resources must be prevented.
 
-## WebGPU Execution Environment Specification for SPIR-V
+Validating accesses are in bounds at shader module or pipeline creation time isn't possible because it is equivalent to the halting problem.
+Instead a WebGPU implementation should make sure out of bounds accesses don't cause accesses outside of the ExecutionContext.
+This will typically be done by instrumenting the generated code, and because of the performance sensitive nature of it, implementation will have a choice in the behavior of each out of bounds access.
+Similarly to OpenGL's `GL_KHR_robust_buffer` extension, out of bounds accesses must produce one of the following behaviors:
+
+* Be discarded for writes
+* Access any location within the resource for reads and writes
+* Return zero values for reads or (0, 0, 0, X) with X being 0, 1, -1, or extrema for integers, or -0.0, +0.0, -1.0, +1.0 for floating point values
+* Atomics can return undefined values.
 
+The robust resource access behavior of OpenGL is extended to allow for trapping behavior as well:
+
+* Cause a trap that immediately stops execution of the shader, and fills all of the shader stage's output with zeroes.
```

Comment:
Note:  This is doesn't induce a race because shader stage outputs are private to the invocation.
(That's not true for tessellation shaders, I think, but WebGPU doesn't support tessellation.)

Reply:
Done and Ack, WebGPU won't support tesselation at least in the medium term.